src/optical_spi.c

```diff
@@ -2,143 +2,146 @@
+TODO:
+- implement timeout
 */
 
 #include "optical_spi.h"
 
-// Must use volatile variables because we modify them in the ISR, but read them
-// from the blocking function
+// volatile needed?
+volatile uint8_t spi_frame_number = 0;  // number of bytes received from OPTICAL
+volatile uint32_t opt_spi_data = 0;     // Data received from OPTICAL, always right-aligned
 
-// Number of bytes already received from PAY-Optical
-volatile uint8_t opt_spi_num_bytes = 0;
-// Data already received from PAY-Optical, always right-aligned
-volatile uint32_t opt_spi_data = 0;
+// tracking if SPI in progress
+bool spi_in_progress = false;
 
-
-/*
-Initializes the microcontroller to communicate with the PAY-Optical microcontroller over SPI.
-*/
+// set up slave select, reset, and DATA_RDYn pins
 void init_opt_spi(void) {
-    // set behaviour of INT1 to trigger on rising edge
-    EICRA |= (1 << ISC10) | (1 << ISC11);
-    // enable external interrupt 0
-    EIMSK |= (1 << INT1);
-    // enable global interrupts
-    sei();
-
     // Initialize PAY-Optical CS pin
     init_cs(OPT_CS_PIN, &OPT_CS_DDR );
     set_cs_high(OPT_CS_PIN, &OPT_CS_PORT);
 
     // Initialize PAY-Optical reset pin
     init_cs(OPT_RST_PIN, &OPT_RST_DDR);
     set_cs_high(OPT_RST_PIN, &OPT_RST_PORT);
+
+    // Initialize PAY-Optical DATA_RDY pin
+    OPT_DATA_DDR &= ~_BV(OPT_DATA);     // write 0, set as input pin
+    OPT_DATA_PORT |= _BV(OPT_DATA);     // activate input pullup resistor
+
 }
 
-/*
-Resets the PAY-Optical microcontroller.
-*/
+
+// Resets the PAY-Optical microcontroller
 void rst_opt_spi(void) {
     // TODO - check how many cycles is necessary
     set_cs_low(OPT_RST_PIN, &OPT_RST_PORT);
     _delay_ms(1000);
     set_cs_high(OPT_RST_PIN, &OPT_RST_PORT);
 }
 
-// Reads 24 bits of raw data from the PAY-Optical microcontroller over optical SPI
-// (Sends the request command, then blocks until it gets 3 bytes back)
-uint32_t read_opt_spi(uint8_t field_num) {
-    // Must make this block non-atomic to ensure interrupts are enabled, in case
-    // this function executes within an atomic block (e.g. in main() of main_test)
-    NONATOMIC_BLOCK(NONATOMIC_RESTORESTATE) {
-        send_opt_spi_cmd(field_num);
-
-        // Wait until we get 3 bytes, with a timeout
-        // In testing, timeout always ended at the original value minus 7 or 8
-        uint32_t timeout = UINT32_MAX;
-        while ((timeout > 0) && (opt_spi_num_bytes < 3)) {
-            timeout--;
-        }
-        print("timeout = %u\n", timeout);
-
-        // If we didn't get 3 bytes
-        if (opt_spi_num_bytes < 3) {
-            //_delay_ms(5000); // long delay = timeout, used for testing
-            return 0;
-        }
 
-        // If we have received all 3 bytes now
-        uint32_t read_data = opt_spi_data;
-        // print ("Received optical data: 0x%06lx\n", read_data);
+// returns status of DATA_RDY pin
+uint8_t get_data_pin(void){
+    return OPT_DATA_PIN >> OPT_DATA;
```

Comment:
Need to bitwise AND this with 0x1 after to isolate the one bit (`(OPT_DATA_PIN >> OPT_DATA) & 0x1`)